File: C++/Arrays/medium/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> R;
        for(int i = 0; i < nums.size()-2 ; i++){
            if( i > 0 && nums[i] == nums[i-1]){
                continue;
            }
            int l,r;
            l = i + 1;
            r = nums.size()-1;
            while(l < r){
                if(nums[i] + nums[l] + nums[r] == 0){
                    R.push_back({nums[i], nums[l] , nums[r]});
                    l++;
                    r--;
                    while(l < r && nums[l] == nums[l-1]){
                        l++;
                    }
                }
                else if(nums[i] + nums[l] + nums[r] < 0){
                    l++;
                }
                else{
                    r--;
                }
            }
        }

        return R;
    }
};

File: Python/Arrays/medium/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        R = []
        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            l,r = i+1,len(nums)-1
            while l < r:
                total = nums[i] + nums[l] + nums[r]
                if  total == 0:
                    R.append([nums[i],nums[l],nums[r]])
                    l+=1 
                    r-=1
                    while l < len(nums) and nums[l] == nums[l-1]:
                        l+=1

                elif total < 0:
                    l += 1 
                
                else:
                    r -= 1 
        return R
        